src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -27,11 +30,96 @@ public ParamsSemanticModel(BicepParamFile bicepParamFile, Compilation? bicepComp
             // name binding is done
             // allow type queries now
             paramsSymbolContext.Unlock();
+            
         }
 
-        public IEnumerable<IDiagnostic> GetDiagnostics()
-            => BicepParamFile.ProgramSyntax.GetParseDiagnostics();
+        /// <summary>
+        /// Gets all the params semantic diagnostics unsorted. Does not include params parser and params lexer diagnostics.
+        /// </summary>
+        /// <returns></returns>
+        private IEnumerable<IDiagnostic> GetAdditionalSemanticDiagnostics()
+        {
+            if (this.BicepCompilation is null)
+            {
+                return Enumerable.Empty<IDiagnostic>();
+            }
+            var requiredParameters = this.BicepCompilation.GetEntrypointSemanticModel().Parameters;
+            var parameterAssignmentSyntax = BicepParamFile.ProgramSyntax.Children.OfType<ParameterAssignmentSyntax>();
+
+            var sortedRequiredParameters =  requiredParameters.OrderBy(x => x.Name);
+            var sortedParameterAssignments = parameterAssignmentSyntax.OrderBy(x => this.ParamBinder.GetSymbolInfo(x)?.Name);
+
+            // get errors relating to missing declarations or assignments
+            AddMissingErrors(sortedRequiredParameters.ToArray(), sortedParameterAssignments.ToArray());
+            // get errors relating to type mismatch of params between Bicep and params files
+            AddTypeMismatchErrors(sortedParameterAssignments);
+
+            return semanticDiagnostics;
+        }
+
+        private void AddMissingErrors(Metadata.ParameterMetadata[] sortedRequiredParameters, ParameterAssignmentSyntax[] sortedParameterAssignments)
```

Comment:
Would love suggestions for a better method name

Reply:
`AddParameterMismatchErrors`?

Reply:
Something like `GetParamAssignmentDiagnostics()` maybe?

Reply:
For some reason GH didn't show me Shenglong's comment when I typed this - I think I prefer his suggestion over mine. Doesn't matter too much!